### Problem 1

In [25]:
import pandas as pd
import mlxtend
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
import math

In [2]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx"
df = pd.read_excel(url)

In [3]:
df.head(5)

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

**Using the apriori module from the MLxtend library, generate Frequent Itemsets for all transactions for the country of France.**

Clean-up:
1. Stripping spaces in the description column
2. Dropping rows that doesn't contain invoice numbers
3. Remove credit transactions 

In [4]:
df['Description'] = df['Description'].str.strip()
df.dropna(axis = 0, subset=['InvoiceNo'], inplace = True)
df['InvoiceNo'] = df['InvoiceNo'].astype('str')
df = df[~df['InvoiceNo'].str.contains('C')]

In [5]:
df.groupby('Country').count().reset_index().sort_values('InvoiceNo', ascending = False).head()

Country  InvoiceNo  StockCode  ...  InvoiceDate  UnitPrice  CustomerID
36  United Kingdom     487622     487622  ...       487622     487622      354345
14         Germany       9042       9042  ...         9042       9042        9042
13          France       8408       8408  ...         8408       8408        8342
10            EIRE       7894       7894  ...         7894       7894        7238
31           Spain       2485       2485  ...         2485       2485        2485

[5 rows x 8 columns]

**1-hot encoding :** This is the process pf consolidating items into one transaction per row. 
The colums represent unique items present in the input array, and rows represent the individual transactions. 

In [6]:
Basket = (df[df['Country']=="France"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

Basket.head()

Description  10 COLOUR SPACEBOY PEN  ...  ZINC T-LIGHT HOLDER STARS SMALL
InvoiceNo                            ...                                 
536370                          0.0  ...                              0.0
536852                          0.0  ...                              0.0
536974                          0.0  ...                              0.0
537065                          0.0  ...                              0.0
537463                          0.0  ...                              0.0

[5 rows x 1563 columns]

In-order to complete the one-hot encoding process, we need to replace all values of quantity >=1 by 1. 

In [7]:
def sum_to_boolean(x):
    if x<=0:
        return 0
    else:
        return 1

Basket_Final = Basket.applymap(sum_to_boolean)

In [8]:
Basket_Final.drop('POSTAGE', inplace=True, axis=1)

Basket_Final.head()

Description  10 COLOUR SPACEBOY PEN  ...  ZINC T-LIGHT HOLDER STARS SMALL
InvoiceNo                            ...                                 
536370                            0  ...                                0
536852                            0  ...                                0
536974                            0  ...                                0
537065                            0  ...                                0
537463                            0  ...                                0

[5 rows x 1562 columns]

In [9]:
## Apriori to select the most important itemsets
Frequent_itemsets = apriori(Basket_Final, min_support = 0.05, use_colnames = True)

Frequent_itemsets.sort_values('support', ascending = False).head()

support                              itemsets
46  0.188776                  (RABBIT NIGHT LIGHT)
52  0.181122       (RED TOADSTOOL LED NIGHT LIGHT)
44  0.170918    (PLASTERS IN TIN WOODLAND ANIMALS)
40  0.168367       (PLASTERS IN TIN CIRCUS PARADE)
59  0.158163  (ROUND SNACK BOXES SET OF4 WOODLAND)

*What itemset has the largest support?*

(RABBIT NIGHT LIGHT)

**Set the minimum support threshold to 5% and extract frequent itemsets, and use them as input to the association rules module. Use each of the confidence and lift metrics to extract the association rules with the highest values, respectively.**

In [14]:
Asso_Rules_confidence = association_rules(Frequent_itemsets, metric ="confidence", min_threshold = 0.1)
Asso_Rules_confidence.sort_values('confidence',ascending = False)

antecedents  ... conviction
82  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...  ...  34.489796
80  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...  ...  34.897959
67                    (SET/6 RED SPOTTY PAPER PLATES)  ...  21.556122
8                        (CHILDRENS CUTLERY SPACEBOY)  ...  12.535714
38                     (PACK OF 6 SKULL PAPER PLATES)  ...  10.298469
..                                                ...  ...        ...
49                    (PLASTERS IN TIN CIRCUS PARADE)  ...   1.174911
55                    (RED TOADSTOOL LED NIGHT LIGHT)  ...   1.177296
56                    (RED TOADSTOOL LED NIGHT LIGHT)  ...   1.151939
57                               (RABBIT NIGHT LIGHT)  ...   1.143338
48                    (RED TOADSTOOL LED NIGHT LIGHT)  ...   1.157763

[86 rows x 9 columns]

In [15]:
Asso_Rules_lift = association_rules(Frequent_itemsets, metric = "lift", min_threshold = 1)
Asso_Rules_lift.sort_values('lift',ascending = False)

antecedents  ... conviction
39     (PACK OF 6 SKULL PAPER CUPS)  ...   4.719388
38   (PACK OF 6 SKULL PAPER PLATES)  ...  10.298469
9    (CHILDRENS CUTLERY DOLLY GIRL)  ...   8.690476
8      (CHILDRENS CUTLERY SPACEBOY)  ...  12.535714
71       (ALARM CLOCK BAKELIKE RED)  ...   2.855230
..                              ...  ...        ...
55  (RED TOADSTOOL LED NIGHT LIGHT)  ...   1.177296
48  (RED TOADSTOOL LED NIGHT LIGHT)  ...   1.157763
49  (PLASTERS IN TIN CIRCUS PARADE)  ...   1.174911
57             (RABBIT NIGHT LIGHT)  ...   1.143338
56  (RED TOADSTOOL LED NIGHT LIGHT)  ...   1.151939

[86 rows x 9 columns]

*What are the antecedents and consequents of each rule? Is the rule with the highest confidence the same as the rule with the highest lift? Why or why not?*

We can see the antedecents and consequents of each rule from Asso_Rules_lift and Asso_Rules_confidence. 

No, the rule with the highest confidence is not the same as the rule with the highest lift. This is because confidence is the ratio of the number of transactions that include all items in the consequent, as well as the antecedent to the number of transactions that include all items in the antecedent or probability that Y(consequent) will be purchased given that X (antedecent) was purchased, while lift is the ratio of confidence to expected confidence or how much our confidence has increased that Y(consequent) will be purchased given that X (antedecent) was purchased. These are 2 different measurements, hence the rules with the highest value for each metric are diferent.


Resources:

UCI Online Retail Example - Apriori & Association Rules.ipynb

https://towardsdatascience.com/association-rules-2-aa9a77241654

### Problem 2

In [16]:
bakery = pd.read_csv('75000-out2-binary.csv')

In [17]:
bakery.head(5)

Transaction Number  Chocolate Cake  ...  Cherry Soda  Single Espresso
0                   1               0  ...            0                0
1                   2               0  ...            0                0
2                   3               0  ...            0                0
3                   4               0  ...            0                0
4                   5               0  ...            0                0

[5 rows x 51 columns]

Calculate the binary correlation coefficient Φ for the Chocolate Coffee and Chocolate Cake items.

In [19]:
item1_name = 'Chocolate Cake'
item2_name = 'Chocolate Coffee'

selection = bakery[[item1_name,
                  item2_name]]

In [20]:
item1_count = selection[item1_name] == 1
item2_count = selection[item2_name] == 1

In [21]:
f = selection.groupby([item1_count,
                       item2_count]).count()
print(f)

                                 Chocolate Cake  Chocolate Coffee
Chocolate Cake Chocolate Coffee                                  
False          False                      65802             65802
               True                        2933              2933
True           False                       2962              2962
               True                        3303              3303


In [22]:
f00 = f[item1_name][0][0]
f01 = f[item1_name][0][1]
f10 = f[item2_name][1][0]
f11 = f[item2_name][1][1]

print(f00)
print(f01)
print(f10)
print(f11)

65802
2933
2962
3303


In [23]:
f1p = f11 + f10
fp1 = f11 + f01
f0p = f01 + f00
fp0 = f10 + f00

print(f1p)
print(fp1)
print(f0p)
print(fp0)

6265
6236
68735
68764


In [27]:
phi = (f11*f00-f01*f10)/math.sqrt(f1p*fp1*f0p*fp0)
print(phi)

0.4855664925278768


In [28]:
phi = (f00*f11-f10*f01)/math.sqrt(f1p*fp1*f0p*fp0)
print(phi)

0.4855664925278768


There is a weak possitive correlation between these two items given the phi coefficient of 0.4856.

*Are these two items symmetric binary variables? Provide supporting calculations.*

Since the formula for phi is symmetrical - swapping antedecent with consequent and vice-versa changes nothing (we can notice that by looking at the formula), these two items are symmetric binary variables. 

*Would the association rules {Chocolate Coffee} =⇒ {Chocolate Cake} have the same value for Φ as {Chocolate Cake} =⇒ {Chocolate Coffee}?*

Yes, both of these rules will have the same value for phi, since the formula for phi is symmetrical - swapping antedecent with consequent and vice-versa changes nothing. The phi value is 0.4856

Resources: Notebook - Association Rules & Analysis.ipynb